File: node-api/src/controller/portfolioCategoryController.js
import * as service from "../services/portfolioCategoryService.js";
import { Router } from "express";
import { getAuthentication } from "../utils/jwt.js";

const endpoints = Router();
const auth = getAuthentication();

endpoints.get("/portfolio-category", async (req, resp) => {
  resp.send(await service.getPortfolioCategories());
});

endpoints.get("/portfolio-category/:id", async (req, resp) => {
  const result = await service.getPortfolioCategoryById(Number(req.params.id));

  if (!result) {
    return resp.status(404).send({
      error: "Category not found",
    });
  }

  resp.send(result);
});

endpoints.post("/portfolio-category", auth, async (req, resp) => {
  resp.status(201).send(await service.savePortfolioCategory(req.body));
});

endpoints.put("/portfolio-category/:id", auth, async (req, resp) => {
  resp.send(
    await service.savePortfolioCategory({
      ...req.body,
      id: Number(req.params.id),
    }),
  );
});

endpoints.delete("/portfolio-category/:id", auth, async (req, resp) => {
  try {
    const id = Number(req.params.id);
    const result = await service.deletePortfolioCategory(id);
    resp.send(result);
  } catch (err) {
    return resp.status(err.status ?? 500).send({
      error: err.message,
    });
  }
});

export default endpoints;

File: spring-api/projeto-fotografico/projeto-fotografico/src/main/java/com/sp/senac/projeto_fotografico/controllers/PortfolioCategoryController.java
package com.sp.senac.projeto_fotografico.controllers;

import com.sp.senac.projeto_fotografico.dto.PortfolioCategoryRequest;
import com.sp.senac.projeto_fotografico.services.PortfolioCategoryService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.Map;

@RestController
@RequestMapping("/portfolio-category")
public class PortfolioCategoryController {

    @Autowired
    private PortfolioCategoryService service;

    @GetMapping
    public ResponseEntity<?> getAll() {
        return ResponseEntity.ok(
                service.getPortfolioCategories()
        );
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getById(
            @PathVariable int id
    ) {
        return service.getPortfolioCategoryById(id)
                .<ResponseEntity<?>>map(ResponseEntity::ok)
                .orElse(
                        ResponseEntity.status(404)
                                .body(Map.of("error", "Category not found"))
                );
    }

    @PostMapping
    public ResponseEntity<?> create(
            @RequestBody PortfolioCategoryRequest request
    ) {
        try {
            service.create(request);
            return ResponseEntity.status(201)
                    .body(Map.of("result", "Saved"));
        } catch (Exception e) {
            return ResponseEntity.badRequest()
                    .body(Map.of("error", e.getMessage()));
        }
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> update(
            @PathVariable int id,
            @RequestBody PortfolioCategoryRequest request
    ) {
        try {
            service.update(id, request);
            return ResponseEntity.ok(Map.of("result", "Updated"));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.status(404)
                    .body(Map.of("error", e.getMessage()));
        } catch (Exception e) {
            return ResponseEntity.badRequest()
                    .body(Map.of("error", e.getMessage()));
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> delete(
            @PathVariable int id
    ) {
        try {
            service.delete(id);
            return ResponseEntity.ok(Map.of("success", true));
        } catch (Exception e) {
            return ResponseEntity.status(404)
                    .body(Map.of("error", e.getMessage()));
        }
    }

    @PostMapping("/{id}/view")
    public ResponseEntity<?> incrementView(
            @PathVariable int id
    ) {
        try {
            service.incrementViews(id);
            return ResponseEntity.ok(Map.of("success", true));
        } catch (Exception e) {
            return ResponseEntity.status(404)
                    .body(Map.of("error", e.getMessage()));
        }
    }
}